### PR TITLE
Add WebGPU SPIR-V Assembler in JavaScript.

### PR DESCRIPTION
**Pull request by dj2**
This CL adds a simple JavaScript library which will assemble SPIR-V
Assembly compatable with WebGPU.

### CODE REVIEW

File: tools/sva/src/parser.js [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+  }
+
+  extractOperand(ast, result_type, data) {
+    let t = this.next();
+
+    let name = undefined;
+    let kind = undefined;
+    let value = undefined;
+    let params = [];
+    if (data.kind === "IdResult" || data.kind === "IdRef"
```
Comment (dneto0) [inline tools/sva/src/parser.js]: I think you'll need more cases, as in https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/text.cpp#L210 

There is also memory-semantics-ID, scope-id,  optional-id.

File: tools/sva/src/parser_test.js [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+      assert.equal(31, inst.operand(3).value());
+      assert.equal("Sqrt", inst.operand(3).name());
+      assert.equal(43, inst.operand(4).value());
+      assert.equal("43", inst.operand(4).name());
+    });
+  });
+
+  it.skip("handles spec constant ops", () => {
+    // let input = "%sum = OpSpecConstantOp %i32 IAdd %a %b";
+  });
```
Comment (dneto0) [inline tools/sva/src/parser_test.js]: Please add this stress test from OpCopyMemory. It uses two optional memory access operands, each of which has its own mixed set of enums, some of which with its own parameters. 

https://github.com/KhronosGroup/SPIRV-Tools/blob/master/test/text_to_binary.memory_test.cpp#L274

Here's the whole test, showing you exactly which words are produced.  You can figure out how the mapping goes by staring at it...

```
  std::string spirv =
      "OpCopyMemory %1 %2 Volatile|Nontemporal|"
      "MakePointerVisible %3 "
      "Aligned|MakePointerAvailable|NonPrivatePointer 16 %4\n";
  EXPECT_THAT(CompiledInstructions(spirv),
              Eq(MakeInstruction(SpvOpCopyMemory, {1, 2, 21, 3, 42, 16, 4})));
  std::string disassembly =
      EncodeAndDecodeSuccessfully(spirv, SPV_BINARY_TO_TEXT_OPTION_NONE);
  EXPECT_THAT(disassembly, Eq(spirv));
```

File: tools/sva/src/assembler_test.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    let a = new Assembler(ast);
+    let res = a.assemble();
+
+    assert.lengthOf(res, 12);
+    assert.equal((4 /* word count */ << 16) | 43 /* opcode */, res[8]);
+    assert.equal(1 /* %float */, res[9]);
+    assert.equal(2 /* %float */, res[10]);
+    assert.equal(0x3ecccccd /* 0.400000006 */, res[11]);
+  });
```
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/assembler_test.js]: I'll add a test, the strings assembled through the simple.spv_asm file, so they do work in general.

File: tools/sva/src/ast.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    let next = this.next_id_;
+    this.assigned_ids_[name] = next;
+
+    this.next_id_ += 1;
+    return next;
+  }
+
+  getIdBounds() { return this.next_id_ - 1; }
```
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/ast.js]: When you say strict upper bound, you mean it's the ID last used, or one more than the maximum ID used? If it's the last ID, then -1 is right, if it's one more then the maximum then it's off by one.

File: tools/sva/src/ast.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    return size;
+  }
+
+  type() { return this.type_; }
+
+  rawValue() { return this.value_; }
+
+  value() {
+    if (this.value_ === undefined) {
```
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/ast.js]: This is a bit of a hack. We don't call this value() method until after the while file is parsed and the AST has seen the maximum bound.

File: tools/sva/src/lexer_test.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    let t = l.next();
+    assert.equal(t.type, TokenType.kEOF);
+  });
+
+  describe("strings", () => {
+    it("matches strings", () => {
+      let input = "\"GLSL.std.450\" \"GL\\\"embed\\\"quote\"";
+      let results = ["GLSL.std.450", "GL\\\"embed\\\"quote"];
```
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/lexer_test.js]: What is the right answer for quotes on the inside? I'm outputting \", essentially.

File: tools/sva/src/spirv.data.js [hunk 7]
```diff
@@ -0,0 +1,4567 @@
...
+        "Aligned": {
+          "value": 2,
+          "params": [
+            "LiteralInteger"
+          ]
+        },
+        "Nontemporal": {
+          "value": 4
+        },
+        "MakePointerAvailable": {
```
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/spirv.data.js]: The ASM can have either, so I need both. If they say MakePointerAvailable I have to map that, if they say MakePointerAvailableKHR I also have to be able to map that.

File: tools/sva/src/parser_test.js (hunk 2 above)
Comment (dj2) [reply tools/sva/src/parser_test.js]: Ugh, gross. That's going to fail, heh.

File: tools/sva/src/lexer.js [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+    // Must have at least 2 chars for a string.
+    if (this.cur_pos_ >= this.len_ || (this.cur_pos_ + 1 >= this.len_))
+      return undefined;
+    if (!this.is("\""))
+      return undefined;
+
+    this.cur_pos_ += 1;
+    while (this.cur_pos_ <= this.len_) {
+      if (this.is("\"")) {
```
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/lexer.js]: done.

File: tools/sva/src/lexer_test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (dj2) [reply tools/sva/src/lexer_test.js]: done.

File: tools/sva/src/parser.js (hunk 1 above)
Comment (dj2) [reply tools/sva/src/parser.js]: Will flush this out post land as I create test cases.

File: tools/sva/src/parser_test.js (hunk 2 above)
Comment (dj2) [reply tools/sva/src/parser_test.js]: Added a skip test, will work out post land.

File: tools/sva/src/ast.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (dneto0) [reply tools/sva/src/ast.js]: Right.  Which is what a comment should say.

File: tools/sva/src/spirv.data.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (dneto0) [reply tools/sva/src/spirv.data.js]: Right.(!)

File: tools/sva/src/parser_test.js (hunk 2 above)
Comment (dneto0) [reply tools/sva/src/parser_test.js]: yeah spir-v syntax is deliberately LR parsing. Need to keep a stack of expected tokens, and reading the mask makes you push stack of more expected tokens.

Comment (dneto0) [reply tools/sva/src/parser_test.js]: See https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/text.cpp#L559

File: tools/sva/src/parser.js (hunk 1 above)
Comment (dj2) [reply tools/sva/src/parser.js]: done.

File: tools/sva/src/parser_test.js (hunk 2 above)
Comment (dj2) [reply tools/sva/src/parser_test.js]: ack.

**Pull request merged**